providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java

```diff
@@ -102,43 +105,54 @@ public void shutdown() throws Exception {
         }
     }
 
-    /**
-     * Resets internal state for retrying establishing the stream.
-     */
-    public void resetRetryConnection() {
-        this.eventStreamAttempt = 1;
-        this.eventStreamRetryBackoff = this.startEventStreamRetryBackoff;
-    }
-
     /**
      * Provide the object that can be used to resolve Feature Flag values.
+     *
      * @return a {@link ServiceGrpc.ServiceBlockingStub} for running FF resolution.
      */
     public ServiceGrpc.ServiceBlockingStub getResolver() {
         return serviceBlockingStub.withDeadlineAfter(this.deadline, TimeUnit.MILLISECONDS);
     }
 
-    private void startEventStream() {
-        StreamObserver<Schema.EventStreamResponse> responseObserver =
-                new EventStreamObserver(this.cache, this.stateConsumer, this::restartEventStream);
-        this.serviceStub
-                .eventStream(Schema.EventStreamRequest.getDefaultInstance(), responseObserver);
-    }
+    private void observeEventStream() {
+        while (this.eventStreamAttempt <= this.maxEventStreamRetries) {
+            StreamObserver<Schema.EventStreamResponse> responseObserver =
+                    new EventStreamObserver(sync, this.cache, this::grpcStateConsumer);
+            this.serviceStub.eventStream(Schema.EventStreamRequest.getDefaultInstance(), responseObserver);
+
+            try {
+                synchronized (sync) {
+                    sync.wait();
```

Comment:
thanks, what about the case of handling notify from onCompleted, is this a valid scenario ? as I see in this case, wait() returns without exception, the code then proceeds to sleep() - not sure it this will cause immediate interrupted exception  and return or will finish the sleep and will proceed to unexpected error.
Maybe add log info/debug and return immediately after wait() can make it clear ? what do you think ?
If this specific case is covered by unit test it can be helpful.

Reply:
`onCompleted` means a successful stream completion. If flagd provide stream closure without error, then this will handle this scenario. 

We need to handle this as this is also a final call similar to `onError`

When this (either `onError` or `onComplete`) happens, we disconnect from flagd. This can be resulted from a restart (ex:- new version rollout), OOMKILL or similar. From SDK point of view, we attempt to reconnect with backoff. This is why we have a sleep and continue with a new stream watcher connection

Reply:
I will try to add some unit tests to cover this scenario :)

Reply:
Added unit tests to verify retry attempts - https://github.com/open-feature/java-sdk-contrib/pull/398/commits/1af6f4ce52f326bb12482da906957013acade150#diff-0dff792882d8d7962486094010e2c59c4210efa38a4b0f7ee4d5f964934eb3c4